Extension/src/LanguageServer/configurations.ts

```diff
@@ -739,29 +739,52 @@ export class CppProperties {
-            });
-        }
-        if (!glob) {
-            return result;
+            if (paths && !glob) {
+                paths.forEach(entry => {
+                    const entries: string[] = util.resolveVariables(entry, env).split(util.envDelimiter).map(e => this.resolvePath(e, false)).filter(e => e);
+                    result.push(...entries);
+                });
+                return result;
+            }
         }
 
         const globResult: string[] = [];
-        for (let res of result) {
-            // fastGlob will expand the ending double wildcard. temporary strip them before expanding
-            const recursive: boolean = res.endsWith('**');
-            if (recursive) {
-                res = res.slice(0, res.length - 2);
-            }
-            // fastGlob can't deal with backslash-separated path => remove them
-            const normalized: string = res.replace(/\\/g, '/');
-            const cwd: string = this.rootUri?.fsPath?.replace(/\\/g, '/') || '';
-            // fastGlob silently strip non-found paths. limit that behavior to dynamic paths only
-            const matches: string[] = fastGlob.isDynamicPattern(normalized) ?
-                fastGlob.sync(normalized, { onlyDirectories: true, cwd}) : [res];
-            globResult.push(...matches.map(s => recursive ? s + '**' : s));
+        let recursiveDoubleStar: boolean = false;
+        let recursiveForwardSlash: boolean = false;
+        let recursiveForwardSlashSingleStar: boolean = false;
+        let recursiveDoubleBackwardSlashDoubleStar: boolean = false;
+        if (paths) {
+            for (let res of paths) {
+                recursiveDoubleStar = res.endsWith('**') && res.endsWith('\\**');
+                recursiveForwardSlash = res.endsWith('/');
+                recursiveForwardSlashSingleStar = res.endsWith('/*');
+                recursiveDoubleBackwardSlashDoubleStar = res.endsWith('\\**');
+                // fastGlob will expand the ending double wildcard. temporary strip them before expanding
+                if (recursiveDoubleStar || recursiveForwardSlashSingleStar) {
+                    res = res.slice(0, res.length - 2);
+                } else if (recursiveForwardSlash) {
+                    res = res.slice(0, res.length - 1);
+                } else if (recursiveDoubleBackwardSlashDoubleStar) {
+                    res = res.slice(0, res.length - 4);
+                }
+
+                // fastGlob can't deal with backslash-separated path => remove them
+                const normalized: string = res.replace(/\\/g, '/');
+                const cwd: string = this.rootUri?.fsPath?.replace(/\\/g, '/') || '';
+                // fastGlob silently strip non-found paths. limit that behavior to dynamic paths only
+                const matches: string[] = fastGlob.isDynamicPattern(normalized) ?
+                    fastGlob.sync(normalized, { onlyDirectories: true, cwd }) : [res];
+
+                if (recursiveForwardSlash) {
```

Comment:
The if condition already checks the value of the variable, the second check isn't needed?